### PR TITLE
Pattern discard not allowed for union case that takes no data

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes https://github.com/dotnet/fsharp/issues/13851

- [x] Put this behind a Lang preview

### CODE REVIEW

**Comment by abelbraaksma** [comment]
Great work! This also fixes the (closed with ‘resolution by design’) issues: #10204 and #1592.

**Review by T-Gro — APPROVED** [review]

**Comment by abonie** [comment]
Does this also cover `function` keyword? If so, it would make sense to add a unit test for that too

**Review by T-Gro — APPROVED** [review]

**Comment by vzarytovskii** [comment]
> > Does this also cover `function` keyword? If so, it would make sense to add a unit test for that too
> 
> Yes, it does, I have added more tests. Thanks for the review :) 

Does it affect a single-case unions when using them as a deconstruct syntax in functions? I.e. was it allowed before and is it now? My guess it shouldn't be affecting it?

**Comment by T-Gro** [comment]
It **was** working before (with _ discard), and was not working before when using named arg.
@edgarfgp  : Can you please test what the **new** behavior is?

IMO it would be more consistent with other features it function arguments also show a warning if empty discard is used, but the opinion is not very strongly held.
image

**Comment by edgarfgp** [comment]
> Does it affect single-case unions when using them as a deconstruct syntax in functions? I.e. was it allowed before and is it now? My guess it shouldn't be affecting it, since they'll just be aliases?

Can you give an example in code to better illustrate your question :) ? 

```fsharp
type X = X of string
let x: X = X("Vlad")
let myVal =
 function
 | X _ -> ()
```

**Comment by T-Gro** [comment]
> It **was** working before (with _ discard), and was not working before when using named arg. @edgarfgp : Can you please test what the **new** behavior is?
> 
> IMO it would be more consistent with other features it function arguments also show a warning if empty discard is used, but the opinion is not very strongly held. image

FYI the middle ones gives a FS0725 error (good), and it gives a fixing suggestion of replacing this with an underscore.
=> with this PR, would be good to NOT suggest putting in an underscore in case of empty field list. 
(or the suggestion would rather remove the field alltogether, instead of replacing it with an underscore)


image

**Comment by edgarfgp** [comment]
> > It **was** working before (with _ discard), and was not working before when using named arg. @edgarfgp : Can you please test what the **new** behavior is?

@T-Gro   Ok let me add a test with your code sample :) 


> > IMO it would be more consistent with other features it function arguments also show a warning if empty discard is used, but the opinion is not very strongly held. image
> 

Agreed would be good to show the new warning error to make it consistent 

> FYI the middle ones gives a FS0725 error (good), and it gives a fixing suggestion of replacing this with an underscore. => with this PR, would be good to NOT suggest putting in an underscore in case of empty field list. (or the suggestion would rather remove the field alltogether, instead of replacing it with an underscore)
> 
> image

I guess this quick fix is not longer needed and a new/ or existing one suggesting remove the _ will be more useful
I would not be helpful with any VS tooling(quick fixes) as part of this PR 

Any help is appreciated ;)

**Comment by vzarytovskii** [comment]
> I guess this quick fix is not longer needed and a new/ or existing one suggesting remove the _ will be more useful\nI would not be helpful with any VS tooling(quick fixes) as part of this PR

If you could just create a follow-up issue for removing exciting quick fix for this case and adding new one (remove discard) - we'll take care of it.

File: src/Compiler/FSComp.txt [hunk 1]
```diff
@@ -1554,6 +1554,7 @@ featureSelfTypeConstraints,"self type constraints"
 featureRequiredProperties,"support for required properties"
 featureInitProperties,"support for consuming init properties"
 featureLowercaseDUWhenRequireQualifiedAccess,"Allow lowercase DU when RequireQualifiedAccess attribute"
+featureMatchNotAllowedForUnionCaseWithNoData,"Pattern match discard not allowed for union case that takes no data."
```
Comment (psfinaki) [inline src/Compiler/FSComp.txt:1557]:
```suggestion
featureMatchNotAllowedForUnionCaseWithNoData,"Pattern match discard is not allowed for union case that takes no data."
```

**Review by psfinaki — APPROVED** [review]
Hah I was scared to review this due to the number of files in the PR but then realized it's actually the analyzer fixing code in our repo :D 

Thanks, great work - and especially great testing :)

File: src/Compiler/FSComp.txt (hunk 1 above)
Comment (edgarfgp) [reply src/Compiler/FSComp.txt]: Thanks for the suggestion :)

**Comment by edgarfgp** [comment]
> Hah I was scared to review this due to the number of files in the PR but then realized it's actually the analyzer fixing code in our repo :D
> 
> Thanks, great work - and especially great testing :)

Thanks, Happy to help

**Review by psfinaki — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]
Sweet, thanks a ton for it, Edgar, great job as usual!

**Comment by edgarfgp** [comment]
> Sweet, thanks a ton for it, Edgar, great job as usual!

Thanks for the kind words :) . It always a pleasure to help F# to get better.  

I’m already looking for the next issue to contribute

**Comment by psfinaki** [comment]
@edgarfgp here are some [ideas](https://github.com/dotnet/fsharp/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) for you ;)

**Comment by abelbraaksma** [comment]
Awesome that you tackled this long standing issue! 🥳

**Comment by edgarfgp** [comment]
My pleasure @abelbraaksma , @NinoFloris might also like this one as he raised this issue as well 😀

**Comment by edgarfgp** [comment]
> @edgarfgp here are some [ideas](https://github.com/dotnet/fsharp/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) for you ;)

Thanks @psfinaki . I have most of those issues on my radar . I might ask for some technical guidance in some of them

**Comment by psfinaki** [comment]
@edgarfgp sure - absolutely :)

**Pull request merged**